keras_cv/models/object_detection/retina_net/__internal__/layers/prediction_head.py

```diff
@@ -12,11 +12,13 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import tensorflow as tf
 from tensorflow.keras import initializers
 from tensorflow.keras import layers
 
 
-def PredictionHead(output_filters, bias_initializer):
+@tf.keras.utils.register_keras_serializable(package="keras_cv")
+class PredictionHead(layers.Layer):
```

Comment:
So if you use closure-style functions, the layers used are never assigned anywhere on the subclassed models. For them to be tracked they need to be assigned at some point.

For this reason closure-style layers are really for Functional models (where the tracking happens at model construction time). They should not be assigned to a subclassed model. Maybe we could find a way to address this (e.g. by tracking layers called within the `__call__` scope, to be investigated).

Reply:
Agreed. I think we should use the subclass layer as the standard. Having the __call__ to track those layer might means more magic in the keras base layer. There will be more corner case when save/load happens (what's the attribute name to serialize and desalinize those layers.)

Reply:
it is actually quite weird that the training worked for this.  Even worse, training doesn't work AFTER this change... not sure what is going on there.